### PR TITLE
Umfpack amgcl linsolver testing

### PR DESCRIPTION
**Pull request by hnil**
Preliminary pull request for discussion:
1) New working cpr with preliminary implementation of quasiimpes and true impes
2) new csrsolverbackend for umfpack
3) new backend for testing amgcl: in particular different cpr options
4) the csrsolverbackend could be extended to use the old factory including petsc bindings for general matrixes

All of this is up to discussion. Preliminary tests do not get speedup with low tolerance, but hight tolerance may give better results. 
Missing parts: 
- Reuse of stages of amg setup for faster setup times.
- the preliminary scaling of the equations seem to have a bug
- implementation of transpose solves to facilitate adjiont

### CODE REVIEW

File: opm/autodiff/BlackoilAmg.hpp [hunk 1]
```diff
@@ -89,30 +90,35 @@ Dune::OverlappingSchwarzOperator<M,X,Y,T> createOperator(const Dune::Overlapping
...
-            }
-        }
+    if(param.cpr_use_drs_){
+      const auto endi = matrix->end();
+      for (auto i=matrix->begin(); i!=endi; ++i){
+	const BlockVector& bw = weights[i.index()];
+	const auto endj = (*i).end();
+	  for (auto j=(*i).begin(); j!=endj; ++j){  
+	    {
+	      BlockVector bvec(0.0);
```
Comment (blattms) [inline opm/autodiff/BlackoilAmg.hpp]: Probably a matter of taste but I find this loop hard to read. I think you are handcoding
```
BlockVector bvec(0.0);
block.mtv(bw, block);
block[pressureIndex] = bvec;
```

File: opm/autodiff/BlackoilAmg.hpp [hunk 2]
```diff
@@ -124,19 +130,20 @@ scaleMatrixQuasiImpes(const Operator& op, const Communication& comm,
...
-        for ( std::size_t i = 0; i < Block::dimension; i++ )
-        {
-            if ( i != pressureIndex )
-            {
-                block[pressureIndex] += block[i];
-            }
-        }
+    if(param.cpr_use_drs_){
+      for(std::size_t j=0; j < vector.size(); ++j){
+	double val(0.0);
```
Comment (blattms) [inline opm/autodiff/BlackoilAmg.hpp]: Looks like
```
#include <dune/common/dotproduct.hh>
....
block[pressureIndex = Dune::dot(block, bw);
```

File: opm/autodiff/BlackoilAmg.hpp [hunk 3]
```diff
@@ -973,16 +982,18 @@ class BlackoilAmg
...
     }
 private:
     const CPRParameter& param_;
+    const typename TwoLevelMethod::FineDomainType& weights_;
     std::tuple<std::unique_ptr<Matrix>, Operator> scaledMatrixOperator_;
     std::shared_ptr<Smoother> smoother_;
     LevelTransferPolicy levelTransferPolicy_;
     CoarseSolverPolicy coarseSolverPolicy_;
     TwoLevelMethod twoLevelMethod_;
+  //BlockVector weights_;
```
Comment (blattms) [inline opm/autodiff/BlackoilAmg.hpp]: Should be removed.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 4]
```diff
@@ -169,12 +169,6 @@ namespace Opm {
         {
             // compute global sum of number of cells
             global_nc_ = detail::countGlobalCells(grid_);
-            //find rows of matrix corresponding to overlap
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: Why is this removed? I do not see a connection to this PR.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 5]
```diff
@@ -375,16 +371,6 @@ namespace Opm {
             ebosSimulator_.model().linearizer().linearize();
             ebosSimulator_.problem().endIteration();
 
-            auto& ebosJac = ebosSimulator_.model().linearizer().jacobian();
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: same here.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 6]
```diff
@@ -469,179 +455,42 @@ namespace Opm {
         /// Number of linear iterations used in last call to solveJacobianSystem().
         int linearIterationsLastSolve() const
         {
-            return istlSolver().iterations();
-        }
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: and here

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 7]
```diff
@@ -469,179 +455,42 @@ namespace Opm {
...
-
             wellModel().apply(ebosResid);
+            if (param_.matrix_add_well_contributions_) {
+                wellModel().addWellContributions(ebosJac.istlMatrix());
+            }
 
             // set initial guess
             x = 0.0;
 
-            const Mat& actual_mat_for_prec = matrix_for_preconditioner_ ? *matrix_for_preconditioner_.get() : ebosJac.istlMatrix();
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: and here

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 8]
```diff
@@ -469,179 +455,42 @@ namespace Opm {
...
+            perfTimer.start();
+            ebosSolver.prepare(ebosJac.istlMatrix(), ebosResid);
+            linear_solve_setup_time_ = perfTimer.stop();
+            ebosSolver.solve(x);
+       }
 
-        /*!
-           \brief Adapter to turn a matrix into a linear operator.
 
-           Adapts a matrix to the assembled linear operator interface
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: Now I am completely lost. Maybe rebasing would help?

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 9]
```diff
@@ -51,11 +53,112 @@ NEW_PROP_TAG(Scalar);
...
+// Implementation for ISTL-matrix based operator
+//=====================================================================
+
+/*!
+   \brief Adapter to turn a matrix into a linear operator.
+
+   Adapts a matrix to the assembled linear operator interface
+ */
+  template<class M, class X, class Y, class WellModel, bool overlapping,class TypeTag >
+class WellModelMatrixAdapter : public Dune::AssembledLinearOperator<M,X,Y>
```
Comment (blattms) [inline opm/autodiff/ISTLSolverEbos.hpp]: Seems like the code was not removed as guessed above, but moved. Assuming this for all places with "same here", too,

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 10]
```diff
@@ -88,16 +204,140 @@ namespace Opm
...
+        void eraseMatrix() {
+            matrix_for_preconditioner_.reset();
+        }
+
+        void prepare(const Matrix& M, Vector& b) {
+ 	    matrix_.reset(new Matrix(M));
+            rhs_ = &b;
+	    bool matrix_cont_added = EWOMS_GET_PARAM(TypeTag, bool, MatrixAddWellContributions);
+	    
+	    if(matrix_cont_added){
```
Comment (blattms) [inline opm/autodiff/ISTLSolverEbos.hpp]: Why is this controlled by MatrixAddWellContributions?

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 11]
```diff
@@ -381,11 +622,255 @@ namespace Opm
...
+	  bweights /=1000; // given normal desnistyies this scales weights to about 1
+	  weights[index] = bweights;
+	  ++index;
+	}
+	return weights;
+      }
+
+      void scaleEquationsAndVariables(Vector& weights){
+	// loop over primary variables
+	const auto& sol = simulator_.model().solution(0);
```
Comment (blattms) [inline opm/autodiff/ISTLSolverEbos.hpp]: Maybe we should document somewhere what we are doing here in matrix vector notation.

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 12]
```diff
@@ -381,11 +622,255 @@ namespace Opm
...
+	  }	      
+	}
+      }
+      
+      static void multBlocksInMatrix(Matrix& ebosJac,const MatrixBlockType& trans,bool left=true){
+	const int n = ebosJac.N();
+	//const int np = FluidSystem::numPhases;
+	for (int row_index = 0; row_index < n; ++row_index) {
+	  auto& row = ebosJac[row_index];
+	  auto* dataptr = row.getptr();
```
Comment (blattms) [inline opm/autodiff/ISTLSolverEbos.hpp]: Is this a more C-like manner of doing
```
for( auto&& block: row){
  if(left)
    block = block.leftmultiply(trans);
  else
    block = block.rightmultiply(trans);
}
```
Or am I missing the reason why we need to convert to a pointer?

**Comment by hnil** [comment]
Close the the this pullrequest. The main part is in seperate. The amgcl part is still usefull for testing but do not intended to be included at the moment...
